src/Clash/Explicit/Signal/Delayed/Bundle.hs

```diff
@@ -0,0 +1,212 @@
+{-# LANGUAGE DataKinds              #-}
+{-# LANGUAGE DefaultSignatures      #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE KindSignatures         #-}
+{-# LANGUAGE NoImplicitPrelude      #-}
+{-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
+{-# LANGUAGE TypeOperators          #-}
+
+module Clash.Explicit.Signal.Delayed.Bundle (
+  Bundle,
+  Unbundled,
+  bundle,
+  unbundle,
+  ) where
+
+import           Control.Applicative           (liftA2)
+import           GHC.TypeLits                  (KnownNat)
+import           Prelude                       hiding (head, map, tail)
+
+import           Clash.Explicit.Signal.Delayed (DSignal, toSignal,
+                                                unsafeFromSignal)
+import qualified Clash.Signal.Bundle           as B
+
+import           Clash.Signal.Internal         (Domain)
+import           Clash.Sized.BitVector         (Bit, BitVector)
+import           Clash.Sized.Fixed             (Fixed)
+import           Clash.Sized.Index             (Index)
+import           Clash.Sized.RTree             (RTree, lazyT)
+import           Clash.Sized.Signed            (Signed)
+import           Clash.Sized.Unsigned          (Unsigned)
+import           Clash.Sized.Vector            (Vec, lazyV)
+
+import           GHC.TypeLits                  (Nat)
+
+-- | Isomorphism between a 'Clash.Explicit.Signal.Delayed' of a product type
+-- (e.g. a tuple) and a product type of 'Clash.Explicit.Signal.Delayed''s.
+--
+-- Instances of 'Bundle' must satisfy the following laws:
+--
+-- @
+-- 'bundle' . 'unbundle' = 'id'
+-- 'unbundle' . 'bundle' = 'id'
+-- @
+--
+-- By default, 'bundle' and 'unbundle', are defined as the identity, that is,
+-- writing:
+--
+-- @
+-- data D = A | B
+--
+-- instance Bundle D
+-- @
+--
+-- is the same as:
+--
+-- @
+-- data D = A | B
+--
+-- instance Bundle D where
+--   type 'Unbundled'' domain delay D = 'DSignal'' domain d D
+--   'bundle'   _ s = s
+--   'unbundle' _ s = s
+-- @
+--
+class Bundle a where
+  type Unbundled (domain :: Domain) (d :: Nat) a = res | res -> domain d
```

Comment:
I think we want the type family dependency to be: `res -> domain d a`, just like for `Clash.Signal.Bundle.Bundle` the type family dependency is: `res -> domain a`